File: packages/orchestrator-core/src/secret-provider.ts
import {
  AzureCliCredential,
  AzureDeveloperCliCredential,
  ChainedTokenCredential,
  ManagedIdentityCredential,
} from "@azure/identity";
import { SecretClient } from "@azure/keyvault-secrets";

export class SecretProvider {
  private client?: SecretClient;
  private cache = new Map<string, string>();
  private readonly credential = new ChainedTokenCredential(
    new ManagedIdentityCredential(),
    new AzureDeveloperCliCredential(),
    new AzureCliCredential(),
  );

  constructor(private readonly keyVaultUrl?: string) {}

  async get(name: string, directValue?: string): Promise<string | undefined> {
    if (directValue) {
      return directValue;
    }

    if (this.cache.has(name)) {
      return this.cache.get(name);
    }

    if (!this.keyVaultUrl) {
      return undefined;
    }

    if (!this.client) {
      this.client = new SecretClient(this.keyVaultUrl, this.credential);
    }

    let secret;
    try {
      secret = await this.client.getSecret(name);
    } catch (error) {
      const statusCode = (error as { statusCode?: number }).statusCode;
      const code = (error as { code?: string }).code;

      if (statusCode === 404 || code === "SecretNotFound") {
        return undefined;
      }

      throw error;
    }

    if (secret.value) {
      this.cache.set(name, secret.value);
    }

    return secret.value;
  }
}

File: backend/infrastructure/Assimalign.AI.Orchestrator.Infrastructure/Security/SecretProvider.cs
using Assimalign.AI.Orchestrator.Application.Abstractions.Providers;
using Azure.Identity;
using Azure.Security.KeyVault.Secrets;

namespace Assimalign.AI.Orchestrator.Infrastructure.Security;

public sealed class SecretProvider(string? keyVaultUrl) : ISecretProvider
{
    private readonly Dictionary<string, string> cache = new(StringComparer.OrdinalIgnoreCase);
    private readonly ChainedTokenCredential credential = new(
        new ManagedIdentityCredential(ManagedIdentityId.SystemAssigned),
        new AzureDeveloperCliCredential(),
        new AzureCliCredential());

    private SecretClient? client;

    public async Task<string?> GetAsync(
        string secretName,
        string? directValue = null,
        CancellationToken cancellationToken = default)
    {
        if (!string.IsNullOrWhiteSpace(directValue))
        {
            return directValue;
        }

        if (cache.TryGetValue(secretName, out var cached))
        {
            return cached;
        }

        if (string.IsNullOrWhiteSpace(keyVaultUrl))
        {
            return null;
        }

        client ??= new SecretClient(new Uri(keyVaultUrl), credential);

        try
        {
            var response = await client.GetSecretAsync(secretName, cancellationToken: cancellationToken);
            if (!string.IsNullOrWhiteSpace(response.Value.Value))
            {
                cache[secretName] = response.Value.Value;
            }

            return response.Value.Value;
        }
        catch (Azure.RequestFailedException error) when (error.Status == 404 || error.ErrorCode == "SecretNotFound")
        {
            return null;
        }
    }
}
